content/post/stash-v2025.6.30/index.md

```diff
@@ -0,0 +1,138 @@
+
+In environments where multiple `BackupConfigurations` share the same schedule, all corresponding `BackupSessions` may be triggered simultaneously. Without `TaskQueue`, this can overwhelm the system, leading to resource contention and backup failures.
+
+By enforcing a controlled execution flow, `TaskQueue` ensures:
+
+* Ensures the number of active `BackupSessions` never exceeds the defined concurrency limit.
+* It maintains a `Queue` and processes `BackupSessions` in order, one after another or up to the allowed concurrency.
+* Prevents simultaneous backup job creation, resources are used efficiently, reducing resource spikes.
+* Cluster System resources are used efficiently, based on the concurrency limit.
+* Optimizes resource usage across the cluster by distributing backup execution over time.
+* Backups have a higher chance of completing successfully without overloading the cluster.
+* Increases the reliability of backups by minimizing the risk of failure due to resource exhaustion.
+
+#### How to Enable `TaskQueue`?
+
+To enable `TaskQueue`, you can use the `--enable-task-queue` flag during the installation or upgrade of Stash. Here is an example:
+
+```bash
+$ helm upgrade stash oci://ghcr.io/appscode-charts/stash \
+      --- \
+      -- set global.taskQueue.enabled=true \
+      -- set global.taskQueue.maxConcurrentSessions=<max_concurrent_sessions> \
+      --- 
+```
+Here, 
+- `global.taskQueue.enabled` is set to `true` to enable the `TaskQueue` feature.
+- `global.taskQueue.maxConcurrentSessions` is set to define the maximum number of concurrent `BackupSessions` that can be executed at a time.
+
+
+#### How Stash utilize `TaskQueue`?
+
+When this feature enabled, `Stash` uses a separate controller called `TaskQueueController`,
+
+**It works as follows:**
+
+1. Instead of triggering `BackupSessions` directly, the `Stash` operator creates a resource called `PendingTask` for each `BackupSession`.
+2. Each PendingTask contains the actual `BackupSession` and is monitored by the `TaskQueueController`.
+3. The `TaskQueueController` processes these `PendingTask` resources based on the defined maximum concurrency limit.
+
+**Example of `TaskQueue` YAML**
+
+```yaml
+apiVersion: batch.k8s.appscode.com/v1alpha1
+kind: TaskQueue
+metadata:
+  name: appscode-stash-task-queue
+spec:
+  maxConcurrentTasks: 10
+  tasks:
+  - rules:
+      failed: has(self.status.phase) && self.status.phase == 'Failed'
+      inProgress: has(self.status.phase) && self.status.phase == 'Running'
+      success: has(self.status.phase) && self.status.phase == 'Succeeded'
+    type:
+      group: stash.appscode.com
+      kind: BackupSession
+```
+
+Here,
+- `maxConcurrentTasks` is set to `10`, meaning a maximum of 10 `BackupSessions` can be executed concurrently.
+- If you need to reconfigure the `TaskQueue` after enabling it, you can modify the `maxConcurrentTasks` value according to your cluster's capacity.
+
+
+**Example of `PendingTask` YAML**
```

Comment:
You cant just say , "here is something", without any backgroud, like , I have a backupConfig running on my cluster, its schedule has come. not trigger job doing something.